solutions/coci-16-zamjene.mdx

```diff
@@ -0,0 +1,269 @@
+---
+id: coci-16-zamjene
+source: COCI 2016
+title: Zamjene
+author: Kevin Sheng
+---
+
+[Official Analysis (C++)](https://github.com/mostafa-saad/MyCompetitiveProgramming/blob/master/Olympiad/COCI/official/2017/contest2_solutions)
+
+## Explanation
+
+If you've done [Wormhole Sort](http://usaco.org/index.php?page=viewproblem2&cpid=992),
+this problem should look a bit familiar.
+From that problem's editorial, we know that it's possible to sort a component
+iff each positions' *sorted position* is in the same component as itself.
+This motivates a DSU-esque solution (for more info look at the [Guide solution](/problems/usaco-992-wormhole-sort/solution) for Wormhole Sort).
+
+### Hashing
+
+The main issue is addressing type 3 and type 4 queries.
+We need to be able to efficiently check if a component can be sorted, and if it can't,
+we also need to see how many components it can be paired up with so that they can be sorted as a whole.
+
+To do this, we have to check if a component has all the needed numbers for it to be able to be rearranged into sorted order.
+
+For example, say this was our (0-indexed) array:
+$$
+[16, 2, 10, 7, 12, 5, 3, 7, 20, 11]
+$$
+
+A component that consisted of the indexes $0$, $3$, and $4$ would need the numbers $2$, $7$, and $7$ to be sortable.
+
+We can get all the needed numbers by sorting the given array and getting the numbers at the positions of the components, but how can we quickly check if the two sets of numbers are the same? The answer is hashing.
+
+If we assign each number a certain power of a prime $P$ ($42$ gets a power of $30$, $50$ gets a power of $35$, etc.),
+we can obtain our hash by multiplying each power by the number of times it occurs in the set.
+
+For example, if our component was the multiset ${3, 3, 4, 4, 5}$, and each number's power was itself, our hash would be:
+$$
+2 \cdot P^3 + 2 \cdot P^4 + 1 \cdot P^5
+$$
+
+<Info>
+
+It's also possible to assign each distinct number in the array a random value from a large range like $[1,10^6]$.
```

Comment:
what I actually meant was that if you replace every element of `elem_pow` with a random 64-bit integer the solution should still work.

Reply:
would be good to think about what the collision probability is if you do this

Reply:
> what I actually meant was that if you replace every element of `elem_pow` with a random 64-bit integer the solution should still work.

that's what i meant by this?

Reply:
"from a large range like $[1,10^6]$" is not the same as "a random 64-bit integer"